File: cronjobs/github.py
# Copyright Tharsis Labs Ltd.(Evmos)
# SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/backend/blob/main/LICENSE)

import requests
import base64
import json
import os
from redis_functions import getChains, getTokens, setChains, setTokens


ENVIRONMENT = os.getenv("ENVIRONMENT")

CHAINS_DIRECTORY_URL = "https://api.github.com/repos/evmos/chain-token-registry/git/trees/production?recursive=1" if ENVIRONMENT == "production" else "https://api.github.com/repos/evmos/chain-token-registry/git/trees/main?recursive=1"

PAT = os.getenv("GITHUB_KEY")
if not PAT:
    raise Exception("GITHUB_KEY environment variable must be defined")


def get_git_directory_content(path: str):
    content = []
    headers = {"authorization": "token {0}".format(PAT)}
    try:
        res = requests.get(CHAINS_DIRECTORY_URL, headers=headers, timeout=5)
        parsed = res.json()
        for entry in parsed.get('tree', []):
            if (path in entry.get('path') and 'tree' not in entry.get('type')):
                res = requests.get(entry.get('url'), headers=headers, timeout=5)
                token_detail = res.json()
                file_content = token_detail.get('content')
                if file_content is not None:
                    decoded_token_detail = base64.b64decode(file_content)
                    json_token_detail = json.loads(decoded_token_detail)
                    content.append(json_token_detail)
    except Exception as e:
        print(e)
    return content


def get_chain_config():
    data = None
    data = getChains()
    if data is None or len(data) <= 0:
        data = get_git_directory_content('chainConfig')
        setChains(data)
    return data


def get_tokens():
    data = None
    data = getTokens()
    if data is None:
        data = get_git_directory_content('tokens')
        setTokens(data)
    return data

File: internal/v1/requester/github.go
// Copyright Tharsis Labs Ltd.(Evmos)
// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/backend/blob/main/LICENSE)

package requester

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"

	"github.com/tharsis/dashboard-backend/internal/v1/db"
)

type Tree struct {
	Path string `json:"path"`
	Mode string `json:"mode"`
	Type string `json:"type"`
	Sha  string `json:"sha"`
	URL  string `json:"url"`
}

type TreeResponse struct {
	Sha       string `json:"sha"`
	URL       string `json:"url"`
	Tree      []Tree `json:"tree"`
	Truncated bool   `json:"truncated"`
}

type Content struct {
	Content string `json:"content"`
	Sha     string `json:"sha"`
}

type File struct {
	Content string
	URL     string
}

func QueryGithubWithCache(url string) (string, error) {
	if val, err := db.RedisGetGithubResponse(url); err == nil {
		return val, nil
	}

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", err
	}

	githubKey := os.Getenv("GITHUB_KEY")

	req.Header.Add("authorization", "token "+githubKey)

	resp, err := Client.Do(req)
	if err != nil {
		if val, err := db.RedisGetHithubFallbackResponse(url); err == nil {
			return val, nil
		}
		return "", err
	}

	if resp.StatusCode != 200 {
		if val, err := db.RedisGetHithubFallbackResponse(url); err == nil {
			return val, nil
		}
		return "", fmt.Errorf("github response status code different from 200: %d", resp.StatusCode)

	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)

	if err != nil || len(string(body)) == 0 {
		if val, err := db.RedisGetHithubFallbackResponse(url); err == nil {
			return val, nil
		}
		return "", err
	}

	// NOTE: converting to string and back to bytes is not the best,
	// but to support everything in redis as a string it's worth it
	bodyString := string(body)

	db.RedisSetGithubResponse(url, bodyString)
	db.RedisSetGithubFallbackResponse(url, bodyString)
	return bodyString, nil
}

func getChainTokenRegistryURL() string {
	env := os.Getenv("ENVIRONMENT")
	if env == "production" {
		return "https://api.github.com/repos/evmos/chain-token-registry/git/trees/production?recursive=1"
	}
	return "https://api.github.com/repos/evmos/chain-token-registry/git/trees/main?recursive=1"
}

func GetValidatorDirectory() ([]File, error) {
	ValidatorsDirectoryURL := "https://api.github.com/repos/evmos/validator-directory/git/trees/main?recursive=1"
	return GetJsonsFromFolder(ValidatorsDirectoryURL, "mainnet")
}

func GetERC20TokensDirectory() ([]File, error) {
	ERC20TokensDirectoryURL := getChainTokenRegistryURL()
	return GetJsonsFromFolder(ERC20TokensDirectoryURL, "tokens")
}

func GetNetworkConfig() ([]File, error) {
	url := getChainTokenRegistryURL()
	return GetJsonsFromFolder(url, "chainConfig")
}

func GetJsonsFromFolder(url string, folder string) ([]File, error) {
	res := []File{}
	apiResp, err := QueryGithubWithCache(url)
	if err != nil {
		return []File{}, err
	}

	var m TreeResponse
	err = json.Unmarshal([]byte(apiResp), &m)
	if err != nil {
		return []File{}, err
	}

	for _, t := range m.Tree {
		if t.Mode == "100644" {
			// Is file
			if strings.HasPrefix(t.Path, folder+"/") {
				fileResponse, err := QueryGithubWithCache(t.URL)
				if err != nil {
					return []File{}, err
				}

				var m Content
				err = json.Unmarshal([]byte(fileResponse), &m)
				if err == nil {
					rawDecodedText, err := base64.StdEncoding.DecodeString(m.Content)
					if err != nil {
						return []File{}, err
					}
					res = append(res, File{Content: string(rawDecodedText), URL: t.Path})
				}
			}
		}
	}
	return res, nil
}
